packages/go_router/example/lib/stateful_nested_navigation.dart

```diff
@@ -0,0 +1,420 @@
+                      secondDetailsPath: '/b/details/2'),
+              routes: <RouteBase>[
+                GoRoute(
+                  path: 'details/:param',
+                  builder: (BuildContext context, GoRouterState state) =>
+                      DetailsScreen(label: 'B', param: state.params['param']),
+                ),
+              ],
+            ),
+          ),
+
+          /// The route branch for the third tab of the bottom navigation bar.
+          ShellRouteBranch(
+            /// Since this route branch has a nested StatefulShellRoute as the
+            /// root route, we need to specify what the default location for the
+            /// branch is.
+            defaultLocation: '/c1',
+            rootRoute: StatefulShellRoute.rootRoutes(
+              /// This bottom tab uses a nested shell, wrapping sub routes in a
+              /// top TabBar. In this case, we're using the `rootRoutes`
+              /// convenience constructor.
+              routes: <GoRoute>[
+                GoRoute(
+                  path: '/c1',
+                  builder: (BuildContext context, GoRouterState state) =>
+                      const TabScreen(label: 'C1', detailsPath: '/c1/details'),
+                  routes: <RouteBase>[
+                    GoRoute(
+                      path: 'details',
+                      builder: (BuildContext context, GoRouterState state) =>
+                          const DetailsScreen(label: 'C1', withScaffold: false),
+                    ),
+                  ],
+                ),
+                GoRoute(
+                  path: '/c2',
+                  builder: (BuildContext context, GoRouterState state) =>
+                      const TabScreen(label: 'C2', detailsPath: '/c2/details'),
+                  routes: <RouteBase>[
+                    GoRoute(
+                      path: 'details',
+                      builder: (BuildContext context, GoRouterState state) =>
+                          const DetailsScreen(label: 'C2', withScaffold: false),
+                    ),
+                  ],
+                ),
+              ],
+              builder: (BuildContext context, GoRouterState state,
+                  Widget ignoredNavigatorContainer) {
+                /// For this nested StatefulShellRoute we are using a custom
+                /// container (TabBarView) for the branch navigators, and thus
+                /// ignoring the default navigator contained passed to the
+                /// builder. Custom implementation can access the branch
+                /// navigators via the StatefulShellRouteState
+                /// (see TabbedRootScreen for details).
+                return const TabbedRootScreen();
+              },
+            ),
+          ),
+        ],
+        builder: (BuildContext context, GoRouterState state,
+            Widget navigatorContainer) {
+          return ScaffoldWithNavBar(body: navigatorContainer);
```

Comment:
How do I control the transition animation when switching between tabs? With ShellRoute you can use a CustomTransitionPage. I wasn't able to use an AnimatedSwitcher here:

```
        builder: (BuildContext context, GoRouterState state,
            Widget navigatorContainer) {
          return AnimatedSwitcher(
            duration: const Duration(milliseconds: 500),
            child: ScaffoldWithNavBar(body: navigatorContainer),
          );
        },
        
```

Reply:
Yes, good question, the sample code used to contain an example on how to use animation when using a `BottomTabBar`, but I removed that when I added the nested TabBar, as I figured that showed how animation could be achieved (using a custom container, i.e. `TabBarView`, for the navigator). But I can add back the sample code I hade before also. 

There are some things one needs to keep in mind when implementing animations for a setup like this though. Simple fade transitions can be achieved by for instance implementing `pageBuilder` for the screens under a StatefulShellRoute. But if you want something more elaborate, you need to use a more custom solution, like for instance using `PageView` or a `TabBarView` as in the example. You should of course be able to implement even more advanced animations, but you need to be aware of some details, for instance: 

* The Navigators must always be part of the widget tree to be able to maintain state.
* There must never be duplicate Navigators on the screen at the same time.

I did actually attempt to implement a basic `AnimatedSwitcher` in the sample code, but ran into problems with the second bullet above, and initially also with the fact that AnimatedSwitcher requires the new child to be different from the old one for an animation to occur. So I'm not sure AnimatedSwitcher can be successfully used for this. 

Perhaps some of these details need to be added to the documentation somewhere.